model_selection.py

```diff
@@ -0,0 +1,121 @@
+            "layers_classifier": 2,
+            "channels_classifier": trial.suggest_int("channels_classifier", 64, 128, step=32),
+            "dropout_classifier": trial.suggest_float("dropout_classifier", 0.1, 0.3),
+            "activation_classifier": "ELU"
+        }
+
+        train_loader, val_loader = get_dataloaders(dataset, hyperparams["batch_size"])
+
+        # Generate the model.
+        model = create_model(
+            in_channels=dataset.num_node_features,
+            out_channels=dataset.num_classes,
+            **hyperparams
+        ).to(device)
+
+        optimizer = torch.optim.Adam(
+            model.parameters(),
+            lr=hyperparams["learning_rate"],
+            weight_decay=hyperparams["weight_decay"],
+        )
+        criterion = torch.nn.CrossEntropyLoss()
+
+        writer = SummaryWriter(f'runs/{trial.study.study_name}/trial{trial.number:03d}')
+
+        trainer = Trainer(model, optimizer, criterion, device, writer=writer)
+        trainer.set_early_stopping(patience=patience, min_epochs=min_epochs, objective="acc")
+        trainer.set_optuna_trial_prunning(trial, objective="acc")
+        trainer.train(train_loader, val_loader, epochs=max_epochs)
+
+        return trainer.get_best("val_acc")
+
+    return objective
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="Hyperparameter optimization")
+    parser.add_argument("--study-name", type=str, default="study5")
+    parser.add_argument("--dataset", type=str, default="PROTEINS", help="dataset")
+    parser.add_argument("--storage", type=str, default="sqlite:///storage.db")
+    parser.add_argument("--max-epochs", type=int, default=400, help="max epochs")
+    parser.add_argument("--min-epochs", type=int, default=50, help="min epochs")
+    parser.add_argument("--patience", type=int, default=50, help="early stopping")
+    parser.add_argument("--n-trials", type=int, default=1000)
+    parser.add_argument("--timeout", type=int, default=None)
+    parser.add_argument("--device", type=str, default="cpu", help="device")
+    parser.add_argument("--n-jobs", type=int, default=1)
+    args = parser.parse_args()
+    return args
+
+
+def main(
+        device,
+        dataset,
+        n_trials,
+        timeout,
+        n_jobs,
+        study_name,
+        storage,
+        max_epochs,
+        min_epochs,
+        patience,
+):
+    dataset = get_dataset(dataset)
+
+    study = optuna.create_study(
+        study_name=study_name,
+        direction="maximize",
+        storage=storage,
+        sampler=optuna.samplers.TPESampler(seed=0),
+        load_if_exists=True,
+    )
+
+    study.optimize(
+        define_objective(dataset, device, max_epochs, min_epochs, patience),
+        n_trials=n_trials,
+        timeout=timeout,
+        n_jobs=n_jobs,
+    )
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    print_args(args)
+    main(**vars(args))
```

Comment:
Move to `main.py`. It should:
1. Arguments:
  - dataset name
  - convolution name
  - pooling name
2. Parse arguments:
  - task: "regression", "single_task_classification", "multi_task_classification"
  - evaluation metric: "MSE", "accuracy", "AUROC"
  - num_labels: int (1 for regression, number of classes for single task classif., number of labels for multi task classif.)
  - map convolution name to convolution class
  - map pooling name to pooling class
  - load splits from disk (lists of indices, list of 2-element tuples); if not present, throw error
3. Testing loop:
  - iterate through train/test splits
  - split train into train and valid
  - optimize model on train and valid
  - 3 times:
    - re-split, get new train/valid split
    - train new model with early stopping on new valid set
    - test new model, append test result to inner test results list
  - get avg test result from inner test results
  - append test result to outer test results
4. Report results (avg and stddev of outer test results)

`split_dataset.py`
argument: dataset name
splits dataset, saves indices on disk

splits:
  1. fold 1: train/test
  2. fold 2: train/test
  ...
  10. fold 10: train/test

Reply:
Implement evaluation functions for different tasks:
```
def evaluate_regression(y_true, y_pred)
```

```
def evaluate_single_task_classification(y_true, y_pred, metric: str):
    metric_fun = ... # map metric name to function
```

```
def evaluate_multi_task_classification(y_true, y_pred, metric: str):
    for each column in y_true and y_pred:
        col_metric = evaluate_single_task_classification(y_true[:, i], y_pred[:, i], metric)
    return avg(column metrics)
```